src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCopyFileFromContainer.groovy

```
@@ -15,55 +15,209 @@
+                hostDestination.delete()
+        } else {
+            hostDestination.parentFile.mkdirs()
         }
-        finally {
-            gzipOut?.close()
+
+        new File(compressedFileLocation,
+                compressedFileName).
+                withOutputStream { it << tarStream }
+
+    }
+
+    /**
+     * Copy regular file or directory from container to host
+     */
+    private void copyFile(InputStream tarStream, File hostDestination) {
+
+        def tempFile
+        def tempDir
+        try {
+
+            // Write tar to temp location since we are exploding it anyway
+            tempFile = File.createTempFile(UUID.randomUUID().toString(), ".tar")
+            tempFile.withOutputStream { it << tarStream }
+
+            // We are not allowed to rename tempDir's created in OS temp directory (as
+            // we do further downstream) which is why we are creating it relative to
+            // build directory
+            tempDir = new File("${project.buildDir}/${UUID.randomUUID().toString()}")
+            tempDir.mkdirs()
+            project.copy {
+                into tempDir
+                from project.tarTree(tempFile)
+            }
+            tempFile.delete()
+
+            /*
+                At this juncture we have 3 possibilities:
+
+                    1.) 0 files were found in which case we do nothing
+
+                    2.) 1 regular file was found
+
+                    3.) N regular files (and possibly directories) were found
+            */
+            def fileCount = 0
+            tempDir.eachFileRecurse(FileType.FILES) { fileCount++ }
+            if (fileCount == 0) {
+
+                logger.quiet "Nothing to copy."
+
+            } else if (fileCount == 1) {
+
+                // ensure regular file does not exist as we don't want clobbering
+                if (hostDestination.exists() && !hostDestination.isDirectory())
+                    hostDestination.delete()
```

We'll need to make sure the return value of `delete()` and `mkdirs()` is handled. Otherwise, we'll swallow an error and the task will fail. Need to check for result and throw `GradleException` if not true.